joinservice/internal/certcache/certcache.go

```diff
@@ -0,0 +1,201 @@
+	kubeClient kubeClient
+}
+
+// NewClient creates a new CertCacheClient.
+func NewClient(log *logger.Logger, kubeClient kubeClient, attVariant variant.Variant) *Client {
+	kdsClient := amdkds.NewKDSClient(trust.DefaultHTTPSGetter())
+
+	return &Client{
+		attVariant: attVariant,
+		log:        log,
+		kubeClient: kubeClient,
+		kdsClient:  kdsClient,
+	}
+}
+
+// CreateCertChainCache creates a certificate chain cache for the given attestation variant
+// and returns the cached certificates, if applicable.
+// If the certificate chain cache already exists, nothing is done.
+func (c *Client) CreateCertChainCache(ctx context.Context) (*CachedCerts, error) {
+	switch c.attVariant {
+	case variant.AzureSEVSNP{}:
+		c.log.Debugf("Creating azure SEV-SNP certificate chain cache")
+		ask, ark, err := c.createCertChainCache(ctx, abi.VcekReportSigner)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create azure SEV-SNP certificate chain cache: %w", err)
+		}
+		return &CachedCerts{
+			ask: ask,
+			ark: ark,
+		}, nil
+	// TODO(derpsteb): Add AWS
+	default:
+		c.log.Debugf("No certificate chain caching possible for attestation variant %s", c.attVariant)
+		return nil, nil
+	}
+}
+
+// CachedCerts contains the cached certificates.
+type CachedCerts struct {
+	ask *x509.Certificate
+	ark *x509.Certificate
+}
+
+// SevSnpCerts returns the cached SEV-SNP ASK and ARK certificates.
+func (c *CachedCerts) SevSnpCerts() (ask, ark *x509.Certificate) {
+	return c.ask, c.ark
+}
+
+// createCertChainCache creates a certificate chain cache configmap with the ASK and ARK
+// retrieved from the KDS and returns ASK and ARK. If the configmap already exists and both ASK and ARK are present,
+// nothing is done and the existing ASK and ARK are returned. If the configmap already exists but either ASK or ARK
+// are missing, the missing certificate is retrieved from the KDS and the configmap is updated with the missing value.
+func (c *Client) createCertChainCache(ctx context.Context, signingType abi.ReportSigner) (ask, ark *x509.Certificate, err error) {
+	c.log.Debugf("Creating certificate chain cache")
+	var shouldCreateConfigMap bool
+
+	cacheAsk, cacheArk, err := c.getCertChainCache(ctx)
+	if k8serrors.IsNotFound(err) {
+		c.log.Debugf("Certificate chain cache does not exist")
+		shouldCreateConfigMap = true
+	} else if err != nil {
+		return nil, nil, fmt.Errorf("failed to get certificate chain cache: %w", err)
+	}
+	if cacheAsk != nil && cacheArk != nil {
+		c.log.Debugf("ASK and ARK present in cache, returning cached values")
+		return cacheAsk, cacheArk, nil
+	}
+	if cacheAsk != nil {
+		ask = cacheAsk
+	}
+	if cacheArk != nil {
+		ark = cacheArk
+	}
+
+	c.log.Debugf("Retrieving certificate chain from KDS")
+	kdsAsk, kdsArk, err := c.kdsClient.CertChain(signingType)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to retrieve certificate chain from KDS: %w", err)
+	}
+	if kdsAsk != nil {
+		ask = kdsAsk
+	}
+	if kdsArk != nil {
+		ark = kdsArk
+	}
```

Comment:
It's not clear if the preference to take certs from KDS than from Cache is wanted. Please clarify, at least with a comment, or with a test case.

Reply:
Sort of. If only the ASK or only the ARK is cached, both will be replaced with the certs from the KDS.